lib/lf/base/comm.cc

```diff
@@ -0,0 +1,180 @@
+#include "comm.h"
+
+namespace lf::base {
+
+namespace comm {
+
+namespace variables {
+
+std::map<std::string, std::pair<bs::hold_any, std::string>> kGlobalVars;
+
+void ListVariables() {
+  for (auto const& el : kGlobalVars) {
+    std::string key = el.first;
+    auto val = el.second;  // nested pairs. Clearer than el.second.second.
+    std::cout << key << " = " << val.first;
+    if (val.second != "")  // if description not empty
+      std::cout << " : " << val.second;
+    std::cout << "\n";
+  }
+}
+
+/**
+ * @brief Checks if key exists in kGlobalVars.
+ * @param key The key of the (key, value) pair.
+ * @return true, if key exists, false otherwise.
+ */
+bool IsSet(const std::string& key) {
+  return kGlobalVars.find(key) != kGlobalVars.end();
+}
+
+/**
+ * @brief Removes key and return true if it existed, false otherwise.
+ * @param key The key of the (key, value) pair.
+ * @return true if key existed, false otherwise.
+ */
+bool Remove(const std::string& key) {
+  auto it = kGlobalVars.find(key);
+  if (it != kGlobalVars.end()) {
+    kGlobalVars.erase(it);
+    return true;
+  }
+  return false;
+}
+
+}  // namespace variables
+
+namespace input {
+
+int kArgc;
+char** kArgv;
+std::string kConfigFile;
+po::variables_map kVM;
+po::options_description kDesc("Allowed options");
+
+/**
+ * @brief Initialises parameters to read from command line (argc, argv) and
+ *        for reading from a file. Also sets default options debug_code and
+ *        debug_levels.
+ * @param argc argc from `int main(int argc, char** argv)`.
+ * @param argv argv from `int main(int argc, char** argv)`.
+ * @param file A file containing options in form name=value.
+ */
+void Init(int argc, char** argv, std::string file) {
+  kArgc = argc;
+  kArgv = argv;
+  kConfigFile = file;
+
+  // check for StaticVars defined
+  // ctrl_root globally defined in static_vars.cc
+  const StaticVar* it = ctrl_root;
+  while (it != nullptr) {
```

Comment:
relying on StaticVar/Track seems a bit of overkill, especially because `Track` uses a rather complicated system to keep track of global variables. A much simpler approach (which would also allow you to get rid of the `Init()` routine completly would be to redefine Track as follows:
```
template<class T>
struct Track{
  Track(std::string name, T& ref, std::string comment = "") {
        try {
          // Don't add the option if it exists already (then no error is thrown)
          // false -> only exact match in name is admissible
          po::option_description el = kDesc.find(it->name_, false);
      } catch (const std::exception e) {
        Add()(name_.c_str(), po::value<T>(&ref),
              comment_.c_str());
    }
  }
};
```

And then define `ADDOPTION` e.g. as follows:
```
#define ADDOPTION(uintvar, name, comment)                             \
  unsigned int uintvar = 0;                                           \
  lf::base::Track<unsigned int> name(#name, uintvar,            \
                                       comment)
```

Reply:
Very nice, I like it!
I implemented this roughly now and will do a cleanup soon.

Reply:
good, let me know when you're done and I'll have a look